theories/datatypes/List.ec

```diff
@@ -1818,6 +1819,137 @@ lemma mapiP x0 (f : int -> 'a -> 'b) (s : 'a list) y :
+(* -------------------------------------------------------------------- *)
+op put (s : 'a list) (n : int) (x : 'a) =
+  if (0 <= n < size s)
+  then take n s ++ (x :: drop (n + 1) s)
+  else s.
+
+lemma put_empty (n : int) (x : 'a) :
+  put [] n x = [].
+proof. by smt(). qed.
+
+lemma size_put (s : 'a list) (n : int) (x : 'a) :
+  size (put s n x) = size s.
+proof.
+rewrite /put; case (0 <= n < size s) => [[? lts_n] | //].
+by rewrite size_cat size_take // lts_n /= size_drop /#.
+qed.
+
+lemma put_in (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => put s n x = take n s ++ (x :: drop (n + 1) s).
+proof. by rewrite /put => ->. qed. 
+
+lemma put_out (s : 'a list) (n : int) (x : 'a) :
+  !(0 <= n < size s) => put s n x = s.
+proof. by rewrite /put => ->. qed. 
+
+lemma put_neg (s : 'a list) (n : int) (x : 'a) :
+  n < 0 => put s n x = s.
+proof. by move => ?; rewrite put_out /#. qed.
+
+lemma put_oversize (s : 'a list) (n : int) (x : 'a) :
+  size s <= n => put s n x = s.
+proof. by move => ?; rewrite put_out /#. qed.
+
+lemma put_cons_0 (s : 'a list) (x y : 'a) :
+  put (y :: s) 0 x = x :: s.
+proof. by rewrite /put /= drop0 addrC ltzS size_ge0. qed. 
+  
+lemma put_cons (s : 'a list) (n : int) (x y : 'a) :
+  n <> 0 => put (y :: s) n x = y :: (put s (n - 1) x).
+proof. by move => ? @/put /#. qed.
+
+lemma put_cat (s1 s2 : 'a list) (n : int) (x : 'a) :
+  put (s1 ++ s2) n x = 
+  if n < size s1 then put s1 n x ++ s2 else s1 ++ put s2 (n - size s1) x.
+proof.
+case (n < 0) => [? |]; first by smt(put_neg size_ge0).
+elim: s1 s2 n => [/#| y s1 ih s2 n ge0_n]; case (n = 0) => [-> | neq0_n].
++ by rewrite ?put_cons_0; smt(size_ge0).
++ by rewrite ?put_cons // ih /#.
+qed.
+
+lemma put_rcons_size (s : 'a list) (x y : 'a) :
+  put (rcons s y) (size s) x = rcons s x.
+proof. 
+rewrite /put size_rcons size_ge0 (_ : size s < size s + 1) 1:/# /=. 
+by rewrite -{1}cats1 drop_oversize 1:size_rcons // take_cat cats0 cats1.
+qed.
+
+lemma put_rcons (s : 'a list) (n : int) (x y : 'a) :
+  n <> size s => put (rcons s y) n x = rcons (put s n x) y.
+proof. 
+move => ?; rewrite -?cats1 put_cat.
+by case (n < size s) => // /lezNgt ?; rewrite ?put_oversize /#.
+qed.
+
+lemma nth_put (dflt : 'a) (s : 'a list) (n : int) (x : 'a) :
+  0 <= n < size s => nth dflt (put s n x) n = x.
+proof.
+by move => [ge0_n lts1_n] @/put; rewrite ge0_n lts1_n /= nth_cat size_take 2:lts1_n.
+qed.
+
+lemma nth_put_out (dflt : 'a) (s : 'a list) (n : int)  (x : 'a) :
+  !(0 <= n < size s) => nth dflt (put s n x) n = dflt.
+proof. by move => ?; rewrite put_out 2:nth_out. qed.
+
+lemma nth_put_neq (dflt : 'a) (s : 'a list) (n1 n2 : int) (x : 'a) :
+  n1 <> n2 => nth dflt (put s n1 x) n2 = nth dflt s n2.
+proof.
+move => ? @/put; case (0 <= n1 < size s) => [[? lts_n1]|] //.
+rewrite nth_cat size_take 2:lts_n1 //= ; case (n2 < n1) => ?.
++ by rewrite nth_take.
++ by rewrite (: n2 - n1 <> 0) /= 2:nth_drop /#.
+qed.
```

Comment:
I think all of this can be replaced:

```suggestion
lemma nth_put x0 (s : 'a list) n1 n2 x :
  0 <= n1 < size s => 
  nth x0 (put s n1 x) n2 = if n1 = n2 then x else nth x0 s n2.
proof. 
move => n1_in; rewrite /put n1_in /= nth_cat size_take 1?(:n1 < size s) /= 1,2:/#.
case (n1 = n2) => [<- //|?]; rewrite (: n2 - n1 <> 0) 1:/# /=.
by case (n2 < n1) => ?; rewrite ?nth_take ?nth_drop /#.
qed.
```

Reply:
Sure, that makes sense. Question: do you actually prefer not explicitly denoting the types of parameters when EC can infer them? I have personally always liked the types being written down explicitly, but I could change this if the convention in the standard library is to leave the types implicit and let EC infer them.